File: client/src/main/java/com/jraska/github/client/push/RemoteMessageToActionConverter.java
package com.jraska.github.client.push;

import com.google.firebase.messaging.RemoteMessage;
import com.jraska.github.client.common.Preconditions;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

final class RemoteMessageToActionConverter {
  public static PushAction convert(RemoteMessage remoteMessage) {
    return convert(remoteMessage.getData());
  }

  public static PushAction convert(Map<String, String> remoteMessageData) {
    Preconditions.argNotNull(remoteMessageData);

    String action = remoteMessageData.get(PushAction.KEY_ACTION);
    if (action == null) {
      return PushAction.DEFAULT;
    }

    Map<String, String> properties;
    if (remoteMessageData.size() == 1) {
      properties = Collections.emptyMap();
    } else {
      properties = new HashMap<>(remoteMessageData);
      properties.remove(PushAction.KEY_ACTION);
    }

    return PushAction.create(action, properties);
  }
}

File: client/src/main/java/com/jraska/github/client/push/RemoteMessageToActionConverter.kt
package com.jraska.github.client.push

import com.google.firebase.messaging.RemoteMessage
import java.util.*

internal object RemoteMessageToActionConverter {
  fun convert(remoteMessage: RemoteMessage): PushAction {
    return convert(remoteMessage.data)
  }

  fun convert(remoteMessageData: Map<String, String>): PushAction {
    val action = remoteMessageData[PushAction.KEY_ACTION] ?: return PushAction.DEFAULT

    val properties: Map<String, String>
    if (remoteMessageData.size == 1) {
      properties = emptyMap()
    } else {
      properties = HashMap(remoteMessageData)
      properties.remove(PushAction.KEY_ACTION)
    }

    return PushAction.create(action, properties)
  }
}
